src/classes/RD2_RecurringDonations_TDTM.cls

```diff
@@ -81,30 +86,40 @@ public class RD2_RecurringDonations_TDTM extends TDTM_Runnable {
         List<SObject> rds, List<SObject> oldRds,
         TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult
     ) {
+        
+        this.currentTriggerAction = triggerAction;
 
         if (!RD2_EnablementService.isRecurringDonations2Enabled) {
             return null;
         }
         if (TDTM_ProcessControl.isRecursive(TDTM_ProcessControl.flag.RD)) {
             return null;
         }
-        if (triggerAction != TDTM_Runnable.Action.BeforeInsert && triggerAction != TDTM_Runnable.Action.BeforeUpdate) {
+        if (currentTriggerAction != TDTM_Runnable.Action.BeforeInsert && currentTriggerAction != TDTM_Runnable.Action.BeforeUpdate) {
             TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.RD, true);
         }
 
-        autogenerateNames(rds, oldRds, triggerAction);
-        validate(rds, oldRds, triggerAction);
-        handleRDSchedules(rds, oldRds, triggerAction);
+        autogenerateNames(rds, oldRds);
+
+        assignDefaultInstallmentFrequency(rds);
+
+        Boolean dataIsValid = validate(rds, oldRds);
+
+        if (!dataIsValid) {
```

Comment:
I agree the the class level variable will not work and have to be fixed. The problem right now is at RD2_ScheduleService line 422 `Integer.valueOf(schedule.InstallmentFrequency__c)` If the RD is updated to null, the code will throw an error at this line because it's an null pointer exception error. I will move the class level variable back to per record and add an extra null check in schedule class to prohibit any RD is entered with Installment Frequency being null.

Reply:
There can be calls to the schedule service outside of the RD2 trigger. See if the schedule service needs to be updated with minimal changes to the RD2 trigger or data migration etc.